File: common/src/main/scala/com/taskadapter/core/TaskKeeperLocationStorage.scala
package com.taskadapter.core

import java.io.File

import com.google.common.base.Charsets
import com.google.common.io.Files
import com.taskadapter.connector.definition.TaskId
import io.circe.generic.auto._
import io.circe.parser._
import io.circe.syntax._

import scala.util.Random

case class TaskKeeperLocation(location1: String, location2: String, cacheFileLocation: String)

case class PreviouslyCreatedTasksCache(location1: String, location2: String, items: Seq[(TaskId, TaskId)])

class TaskKeeperLocationStorage(rootFolder: File) {
  val cacheFolder = new File(rootFolder, "cache")
  val fileName = "cache_file_links.json"

  /**
    * Store elements in the cache on disk. New items are added to existing ones, skipping duplicates.
    * If file does not exist yet, it will be created.
    */
  def store(location1: String, location2: String, items: Seq[(TaskId, TaskId)]): Unit = {
    val file = getOrCreateFileLocation(location1, location2)
    val existingCache = loadCache(location1, location2)
    val allItems = (existingCache.items ++ items).distinct
    val newCache = PreviouslyCreatedTasksCache(location1, location2, allItems)
    val json = newCache.asJson
    val formattedJson = json.spaces2
    Files.write(formattedJson, file, Charsets.UTF_8)
  }

  def loadCache(location1: String, location2: String): PreviouslyCreatedTasksCache = {
    val file = getOrCreateFileLocation(location1, location2)
    val fileBody = Files.toString(file, Charsets.UTF_8)
    if (fileBody.isEmpty) {
      PreviouslyCreatedTasksCache("", "", Seq())
    } else {
      val cache = decode[PreviouslyCreatedTasksCache](fileBody)
      cache match {
        case Right(c) => c
        case Left(e) =>
          throw new RuntimeException(s"cannot parse tasks map from file $file: $e")
      }
    }
  }

  def loadTasks(location1: String, location2: String): PreviouslyCreatedTasksResolver = {
    new PreviouslyCreatedTasksResolver(loadCache(location1, location2))
  }

  private def saveCacheLocation(location1: String, location2: String, newFileName: String): Unit = {
    val previousEntries = loadCache()
    val newEntries = previousEntries ++ List(TaskKeeperLocation(location1, location2, newFileName))

    val jsonString = newEntries.asJson.spaces2
    Files.write(jsonString, new File(cacheFolder, fileName), Charsets.UTF_8)
  }

  private def loadCache(): Seq[TaskKeeperLocation] = {
    val file = new File(cacheFolder, fileName)
    if (file.exists()) {
      val jsonString = Files.toString(file, Charsets.UTF_8)
      decode[Seq[TaskKeeperLocation]](jsonString) match {
        case Left(e) => throw new RuntimeException(s"cannot load file with cache definitions: $e")
        case Right(list) => list
      }
    } else {
      List()
    }
  }

  private def getOrCreateFileLocation(location1: String, location2: String): File = {
    val linksToCaches = loadCache()
    val filePath = findCacheLocation(linksToCaches, location1, location2)
    filePath match {
      case None =>
        val newFileName = Math.abs(Random.nextInt()) + ".json"
        val newFile = new File(cacheFolder, newFileName)
        cacheFolder.mkdirs()
        newFile.createNewFile()

        saveCacheLocation(location1, location2, newFileName)
        newFile.getAbsoluteFile
      case Some(p) =>
        new File(cacheFolder, p).getAbsoluteFile
    }
  }

  private def findCacheLocation(caches: Seq[TaskKeeperLocation], location1: String, location2: String): Option[String] = {
    caches.find(e =>
      // either direction is fine
      (e.location1 == location1 && e.location2 == location2)
        ||
        (e.location2 == location1 && e.location1 == location2)
    ).map(_.cacheFileLocation)
  }

}

File: webui/src/main/java/com/taskadapter/web/TaskKeeperLocationStorage.java
package com.taskadapter.web;

import com.google.common.collect.Lists;
import com.google.common.collect.Sets;
import com.google.gson.JsonParser;
import com.google.gson.reflect.TypeToken;
import com.taskadapter.connector.common.ConfigUtils;
import com.taskadapter.common.JsonUtil;
import com.taskadapter.connector.definition.TaskKeyMapping;
import com.taskadapter.core.PreviouslyCreatedTasksCache;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Random;

import com.google.common.io.Files;
import com.taskadapter.core.PreviouslyCreatedTasksResolver;
import com.taskadapter.core.TaskKeeperLocation;

public class TaskKeeperLocationStorage {
    private static final String fileName = "cache_file_links.json";
    private final File cacheFolder;

    public TaskKeeperLocationStorage(File rootFolder) {
        cacheFolder = new File(rootFolder, "cache");
    }

    /**
     * Store elements in the cache on disk. New items are added to existing ones, skipping duplicates.
     * If file does not exist yet, it will be created.
     */
    public void store(String location1, String location2, List<TaskKeyMapping> items) throws IOException {
        var file = getOrCreateFileLocation(location1, location2);
        var existingCache = loadCache(location1, location2);
        var allItems = new ArrayList<>(existingCache.getItems());
        allItems.addAll(items);
        List<TaskKeyMapping> distinctItems = Lists.newArrayList(Sets.newHashSet(allItems));
        var newCache = new PreviouslyCreatedTasksCache(location1, location2, distinctItems);

        var jsonString = JsonUtil.toJsonString(newCache);
        Files.write(jsonString, file, StandardCharsets.UTF_8);
    }

    public PreviouslyCreatedTasksResolver loadTasks(String location1, String location2) throws IOException {
        return new PreviouslyCreatedTasksResolver(loadCache(location1, location2));
    }

    private PreviouslyCreatedTasksCache loadCache(String location1, String location2) throws IOException {
        var file = getOrCreateFileLocation(location1, location2);
        if (!file.exists() || file.length() == 0) {
            return new PreviouslyCreatedTasksCache("", "", java.util.List.of());
        }
        var fileBody = Files.toString(file, StandardCharsets.UTF_8);
        return JsonUtil.parseJsonString(fileBody, PreviouslyCreatedTasksCache.class);
    }

    private void saveCacheLocation(String location1, String location2, String newFileName) throws IOException {
        var previousEntries = loadCache();
        var newEntries = previousEntries;
        newEntries.add(new TaskKeeperLocation(location1, location2, newFileName));

        var jsonString = ConfigUtils.createDefaultGson().toJsonTree(newEntries).toString();
        Files.write(jsonString, new File(cacheFolder, fileName), StandardCharsets.UTF_8);
    }

    private List<TaskKeeperLocation> loadCache() throws IOException {
        var file = new File(cacheFolder, fileName);
        if (file.exists()) {
            var jsonString = Files.toString(file, StandardCharsets.UTF_8);
            var jsonValue = new JsonParser().parse(jsonString);
            List<TaskKeeperLocation> list = ConfigUtils.createDefaultGson().fromJson(jsonValue,
                    new TypeToken<List<TaskKeeperLocation>>() {
                    }.getType());
            return list;
        }
        return new ArrayList<>();
    }

    private File getOrCreateFileLocation(String location1, String location2) throws IOException {
        var linksToCaches = loadCache();
        var filePath = findCacheLocation(linksToCaches, location1, location2);
        if (filePath.isPresent()) {
            return new File(cacheFolder, filePath.get()).getAbsoluteFile();
        }

        var newFileName = Math.abs(new Random().nextInt()) + ".json";
        var newFile = new File(cacheFolder, newFileName);
        cacheFolder.mkdirs();
        newFile.createNewFile();

        saveCacheLocation(location1, location2, newFileName);
        return newFile.getAbsoluteFile();
    }

    private Optional<String> findCacheLocation(List<TaskKeeperLocation> caches, String location1, String location2) {
        return caches.stream().filter(e ->
                // either direction is fine
                (e.getLocation1().equals(location1) && e.getLocation2().equals(location2))
                        ||
                        (e.getLocation2().equals(location1) && e.getLocation1().equals(location2))
        ).findFirst().map(TaskKeeperLocation::getCacheFileLocation);
    }

    public String getCacheFolderAbsoluteName() {
        return cacheFolder.getAbsolutePath();
    }
}
